Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalPipeline.cpp

```diff
@@ -0,0 +1,4398 @@
+    fState.fCurrentDepthStencilState = fDevice.fNoZReadOrWriteStencilState;
+
+    simd_float4x4 projMat = matrix_identity_float4x4;
+
+    /// Set up the transform directly
+    fDevice.SetLocalToWorldMatrix(plate->GetTransform());
+
+    IPushPiggyBacks(material);
+
+    // First, do we have a device ref at this index?
+    plMetalMaterialShaderRef* mRef = (plMetalMaterialShaderRef*)material->GetDeviceRef();
+
+    if (mRef == nullptr) {
+        mRef = new plMetalMaterialShaderRef(material, this);
+        material->SetDeviceRef(mRef);
+    }
+
+    if (!mRef->IsLinked()) {
+        mRef->Link(&fMatRefList);
+    }
+
+    fDevice.SetLocalToWorldMatrix(plate->GetTransform());
+
+    plMetalPlateManager* pm = (plMetalPlateManager*)fPlateMgr;
+
+    plMetalPlatePipelineState             state(&fDevice);
+    plMetalDevice::plMetalLinkedPipeline* linkedPipeline = state.GetRenderPipelineState();
+
+    if (fState.fCurrentPipelineState != linkedPipeline->pipelineState) {
+        fDevice.CurrentRenderCommandEncoder()->setRenderPipelineState(linkedPipeline->pipelineState);
+        fState.fCurrentPipelineState = linkedPipeline->pipelineState;
+    }
+    float alpha = material->GetLayer(0)->GetOpacity();
+    fDevice.CurrentRenderCommandEncoder()->setFragmentBytes(&alpha, sizeof(float), 6);
+    fDevice.CurrentRenderCommandEncoder()->setDepthStencilState(pm->fDepthState);
+    fDevice.CurrentRenderCommandEncoder()->setCullMode(MTL::CullModeNone);
+
+    int            uniformSize = sizeof(VertexUniforms);
+    VertexUniforms uniforms;
+    uniforms.projectionMatrix = projMat;
+    matrix_float4x4 modelMatrix;
+    uniforms.worldToCameraMatrix = modelMatrix;
+    uniforms.uvTransforms[0].UVWSrc = 0;
+    // uniforms.worldToLocalMatrix = fDevice.fMatrixW2L;
+
+    // flip world to camera, it's upside down
+    matrix_float4x4 flip = matrix_identity_float4x4;
+    flip.columns[1][1] = -1.0f;
+
+    // uniforms.worldToCameraMatrix =
+    // uniforms.cameraToWorldMatrix = fDevice.fMatrixC2W;
+    uniforms.localToWorldMatrix = matrix_multiply(flip, fDevice.fMatrixL2W);
+
+    mRef->FastEncodeArguments(fDevice.CurrentRenderCommandEncoder(), &uniforms, 0);
+    // FIXME: Hacking the old texture drawing into the plate path
+    mRef->prepareTextures(fDevice.CurrentRenderCommandEncoder(), 0);
+
+    fDevice.CurrentRenderCommandEncoder()->setVertexBytes(&uniforms, sizeof(VertexUniforms), VertexShaderArgumentFixedFunctionUniforms);
+
+    pm->EncodeDraw(fDevice.CurrentRenderCommandEncoder());
+
+    IPopPiggyBacks();
+}
+
+// Push and pop light sources
+// The DX version would just keep a giant pool of lights
+// that could be claimed by different parts of the pipeline.
+// In Metal, when a part of the pipeline wants to own lights
+// we'll just let them push/pop the current state.
+void plMetalPipeline::PushCurrentLightSources()
+{
+    plMetalLights* lightSources = new plMetalLights();
+    memcpy(lightSources, &fLights, sizeof(plMetalLights));
```

Comment:
Yes - it's a pod type.

Reply:
I think it would be a good idea to add a compile-time check to make sure this is true. That would look something like:
```C++
#include <type_traits>
static_assert(std::is_trivial_v<plMetalLights>, "plMetalLights must be a trivial type!");
```